### PR TITLE
impl(pubsub): add tracing message flush callback impl

### PR DESCRIPTION
**Pull request by alevenberg**
Part of https://github.com/googleapis/google-cloud-cpp/issues/12528

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12862)

### CODE REVIEW

**Comment by alevenberg** [comment]
There's some overlap with https://github.com/googleapis/google-cloud-cpp/pull/12859. I will rebase and fix the merge conflicts as needed

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 1]
```diff
@@ -40,14 +40,24 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
...
 
-// TODO(#12528): Implement functionality for FlushCallback.
-void TracingMessageBatch::FlushCallback() { child_->FlushCallback(); }
+void TracingMessageBatch::FlushCallback() {
+  for (auto& span : batch_sink_spans_) {
+    internal::EndSpan(*span);
+  }
+  batch_sink_spans_.clear();
+  child_->FlushCallback();
+}
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:49]: I think `FlushCallback()` and `SaveMessage()` can be called concurrently by multiple threads?  This code is not thread safe if I am right.

Consider adding a mutex and changing the implementation to something like:

```suggestion
void TracingMessageBatch::FlushCallback() {
  decltype(batch_sink_spans_) spans;
   {
      std::lock_guard<std::mutex> lk(mu_);
      spans_.swap(batch_sink_spans_);
    }
  for (auto& s : spans) internal::EndSpan(*span);
  child_->FlushCallback();
}
```

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 2]
```diff
@@ -40,14 +40,24 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
...
 
 std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>
 TracingMessageBatch::GetSpans() const {
   return message_spans_;
 }
 
+std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>
+TracingMessageBatch::GetBatchSinkSpans() const {
+  return batch_sink_spans_;
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:58]: I assume this is only interested for testing?  If not, you need a lock.  A lock may be less surprising even if it is only used for testing.

File: google/cloud/pubsub/internal/tracing_message_batch.h [hunk 3]
```diff
@@ -38,6 +38,20 @@ class TracingMessageBatch : public MessageBatch {
  public:
   explicit TracingMessageBatch(std::unique_ptr<MessageBatch> child)
       : child_(std::move(child)) {}
+  TracingMessageBatch(
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch.h:41]: Are these constructors just motivated by the tests?  Can we write the tests so they "naturally" leave the class in the state we want?

**Review by coryan — CHANGES_REQUESTED** [review]

File: google/cloud/pubsub/internal/tracing_message_batch.cc (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc]: Done. I added a lock around the push_back in SaveMessage too.

File: google/cloud/pubsub/internal/tracing_message_batch.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc:63]: Yes. Added comments that it is test only.

File: google/cloud/pubsub/internal/tracing_message_batch.h (hunk 3 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.h:42]: The only alternative I can think of is a giant test that calls .SaveMessage then .Flush, then tests .FlushCallback...

I'm not sure if that's better. 

Do you have any ideas about how I could structure the tests?

File: google/cloud/pubsub/internal/tracing_message_batch.cc (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc]: Is there a good way to test this in a unit test?

File: google/cloud/pubsub/internal/tracing_message_batch.h (hunk 3 above)
Comment (coryan) [reply google/cloud/pubsub/internal/tracing_message_batch.h:42]: I would have written the giant test (maybe not that giant by using for loops?).  I am less worried about internal-only APIs like this, but feels a bit odd to test a class by setting a state only reachable via a test-only constructor.

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 4]
```diff
@@ -33,21 +33,36 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
...
+  for (auto& span : spans) internal::EndSpan(*span);
+  child_->FlushCallback();
+}
 
 std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>
 TracingMessageBatch::GetSpans() const {
   return message_spans_;
 }
 
+std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:61]: hint: I think you can define this function as `auto` and save some typing.

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 5]
```diff
@@ -33,21 +33,36 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
   auto active_span = opentelemetry::trace::GetSpan(
       opentelemetry::context::RuntimeContext::GetCurrent());
   active_span->AddEvent("gl-cpp.added_to_batch");
-  message_spans_.push_back(active_span);
+  {
+    std::lock_guard<std::mutex> lk(mu_);
+    message_spans_.push_back(active_span);
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:38]: I am not sure, but maybe `std::move(active_span)`?

**Review by coryan — APPROVED** [review]

File: google/cloud/pubsub/internal/tracing_message_batch.cc (hunk 5 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc:38]: Done.

**Comment by alevenberg** [comment]
Adding a comment for an end to end test in addition to the per function tests.

**Pull request merged**